msmpi tries to access dependent package hdf5 spack_cc

danlipsa opened the pull request:
The full error:
```
==> Installing hdf5-1.14.3-ohtopgtkb3nmwyxxgqdpwjvvnttospev [24/35] ==> No binary for hdf5-1.14.3-ohtopgtkb3nmwyxxgqdpwjvvnttospev found: installing from source ==> Error: AttributeError: module 'spack.pkg.builtin.hdf5' has no attribute 'spack_cc'

The 'hdf5' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure

C:\Users\dan.lipsa\projects\spack\var\spack\repos\builtin\packages\msmpi\package.py:48, in setup_dependent_package:
         45        # be manually supplied to compiler by consuming package
         46        # Note: This is not typical of MPI installations
         47        dependent_module = dependent_spec.package.module
  >>     48        self.spec.mpicc = dependent_module.spack_cc
         49        self.spec.mpicxx = dependent_module.spack_cxx
         50        self.spec.mpifc = dependent_module.spack_fc
         51        self.spec.mpif77 = dependent_module.spack_f77
```

Changes to msmpi by 1fe8e63481 caused this error.



<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

danlipsa commented:
@johnwparent Please review

alalazo reviewed (dismissed):
Can you improve the title? This is not about `hdf5` but about `msmpi` :slightly_smiling_face: 

Also, I think the fix is logically wrong. `spack_cc` is set to the compiler of `msmpi` rather than the compiler of the dependent.

danlipsa commented:
> Can you improve the title? This is not about `hdf5` but about `msmpi` 🙂
Done.

> Also, I think the fix is logically wrong. `spack_cc` is set to the compiler of `msmpi` rather than the compiler of the dependent.

hdf5 compiles correctly with these changes, so spack_cc reaches the dependent as well. I am reverting to code that @johnwparent put in, so he will have more details on how this works.

johnwparent commented:
> Also, I think the fix is logically wrong. `spack_cc` is set to the compiler of `msmpi` rather than the compiler of the dependent.

I agree with @alalazo here, I think this _should_ be pulling in the compiler of the dependent spec (i.e. if a package is using a compiler wrapper, so should msmpi, etc) as it does not vendor its own compiler wrapper (as this change seems to indicate).

If this is the expected use case, it's not clear to me why it was originally failing for Dan ? IIRC I modeled this after mpich so it should work, and was for a while.

danlipsa commented:
> > Also, I think the fix is logically wrong. `spack_cc` is set to the compiler of `msmpi` rather than the compiler of the dependent.
> 
> I agree with @alalazo here, I think this _should_ be pulling in the compiler of the dependent spec (i.e. if a package is using a compiler wrapper, so should msmpi, etc) as it does not vendor its own compiler wrapper (as this change seems to indicate).

Sorry I don't understand this. The error listed at the beginning says that hdf5 does not does not have dependent_module.spack_cc which is normal because, as you say, there is no mpicc in Windows. This is what it is failing.


> If this is the expected use case, it's not clear to me why it was originally failing for Dan ? IIRC I modeled this after mpich so it should work, and was for a while.

What was working I think, and also what it is working for me, it is what I try to revert to.

danlipsa commented:
@alalazo It seems that `dependent_module.spack_cc` is only available in `setup_dependent_build_environment` but not available in `setup_dependent_package`.
I found only another use inside  `setup_dependent_package` in cray-mvapich2 which was added in the same PR. That is also a wrapper to an external package which I cannot test.

danlipsa committed:
Make sure set_package_py_globals is called for dependent_modules

danlipsa commented:
@alalazo Please review. External modules were processed (set_package_py_globals) before nonexternal modules, so dependents(hdf5) for external modules (msmpi) did not have the attributes (spack_cc) set up properly.

danlipsa commented:
@johnwparent Please review

scheibelp commented on lib/spack/spack/build_environment.py:
```diff
@@ -1000,7 +1000,7 @@ def __init__(self, *specs: spack.spec.Spec, context: Context) -> None:
     def set_all_package_py_globals(self):
         """Set the globals in modules of package.py files."""
-        for dspec, flag in chain(self.external, self.nonexternal):
+        for dspec, flag in chain(self.nonexternal, self.external):
```
Alternative summary:

`msmpi`'s package.py does:

```
    def setup_dependent_package(self, module, dependent_spec):
        ...
        self.spec.mpicc = dependent_module.spack_cc
```

Externals are (almost always) leaves, so this PR ensures that `set_package_py_globals` runs on `hdf5` (the parent) before `setup_dependent_package` runs on `msmpi` (the child).

If it's valid in any circumstance for a child's `setup_dependent_package` function to access attributes of the parent (I think it is), then I think a change (like this PR) is needed.

Another way to approach this is to split the loop into two loops: one for `setup_dependent_package`, and one for `set_package_py_globals`; if it did, the externals-before-non-externals order could be preserved (at least for `msmpi`).

No unit tests fail with the rearrangement in this PR, and I don't see a reason to prefer the original order from looking through `set_package_py_globals`, so I'm curious if @haampie has a preference (based on `git blame`, the latest person to edit this section).

alalazo replied:
Looking at the code, I think this is also wrong. I agree with splitting the loop in two, and:
1. First run `set_package_py_globals` on everything
2. Then re-run the loop and call `setup_dependent_package`.

Also, wouldn't it make sense to run the loop as a topo-sort from leaves to root (instead of external vs. non-external)? @haampie

danlipsa commented:
@scheibelp @alalazo Is this ready?

alalazo reviewed (commented):
This one would be fine with me. Can you add a unit test, to avoid regressions?

danlipsa committed:
Test that set_package_py_globals was already called on dependents

install.py::test_installed_upstream_external fails
if the added assert is triggered.

danlipsa commented:
@alalazo @scheibelp How does the added test look? Thanks!

scheibelp commented on lib/spack/spack/build_environment.py:
```diff
@@ -999,16 +999,17 @@ def __init__(self, *specs: spack.spec.Spec, context: Context) -> None:
                     # This includes runtime dependencies, also runtime deps of direct build deps.
                     set_package_py_globals(pkg, context=Context.RUN)
 
+        for dspec, flag in chain(self.external, self.nonexternal):
```
Can you add a comment here explaining the use of the separate loop? The test will enforce this, but IMO it's still useful to mention that we want to set globals on everything before running `setup_dependent...` on anything.

johnwparent replied:
Done.

scheibelp commented on var/spack/repos/builtin.mock/packages/externaltool/package.py:
```diff
@@ -15,3 +15,9 @@ class Externaltool(Package):
 
     depends_on("externalprereq")
+
+    def setup_dependent_package(self, module, dependent_spec):
```
Can you add an explicit test to `test/build_environment`? It should:

* explicitly configure a builtin.mock package as external (I think `test_external_config_env` has a good example of how to set that up)
* ensure that this method ran (this method will fail if there's a problem and it runs, but we also want to make sure it runs)

Also, I'd recommend localizing this method to the test, either using monkey-patching or defining an entirely new package using the `create_test_repo` function (see `concretize_requirements.py` for an example of that).

johnwparent replied:
Added a test using a monkeypatched `setup_dependent_package` in `externaltool`'s dependency, some validation that method is run at the correct time, and some validation it ran at all.

scheibelp reviewed (changes requested):
@johnwparent I have a couple requests for this

johnwparent committed:
Add more tests and docs

johnwparent commented:
@spackbot fix style

danlipsa committed:
[@spackbot] updating style on behalf of danlipsa

johnwparent committed:
Fix broken test

johnwparent commented:
@spackbot fix style

johnwparent commented:
@spackbot fix style

johnwparent commented:
@spackbot fix style

johnwparent commented:
@spackbot fix style

danlipsa committed:
[@spackbot] updating style on behalf of danlipsa

johnwparent committed:
Sandbox env in new test

johnwparent commented:
@spackbot fix style

danlipsa committed:
[@spackbot] updating style on behalf of danlipsa

scheibelp commented on lib/spack/spack/test/build_environment.py:
```diff
@@ -513,6 +513,26 @@ def test_setting_dtags_based_on_config(config_setting, expected_flag, config, mo
 
+def test_module_globals_available_at_setup(
+    monkeypatch, mutable_config, mock_packages, working_env
+):
```
A couple things:

* can this be renamed to `test_module_globals_available_at_setup_dependent_time` (the current name is confusing IMO)
* Can you add a docstring like:

> Spack-built package externaltest depends on an external package "externaltool". externaltool's setup_dependent_pacakage needs access to globals on the dependent

johnwparent replied:
Both requests done and pushed.

scheibelp commented on lib/spack/spack/build_environment.py:
```diff
@@ -1002,16 +1002,21 @@ def __init__(self, *specs: spack.spec.Spec, context: Context) -> None:
+        # access to expected module level definitions such as "spack_cc"
+        for dspec, flag in chain(self.external, self.nonexternal):
             for spec in dspec.dependents():
+                pkg = dspec.package
```
Can you move this just above the loop? IMO it's clearest to instantiate variables that do not depend on the loop outside of it (that's less important for the `if` statement above but I think it also makes sense to move it there).

johnwparent replied:
Done (also for the if statement)

scheibelp reviewed (changes requested):
I have a couple minor requests

@alalazo I think your original concerns are addressed (from https://github.com/spack/spack/pull/44327#pullrequestreview-2072880845)

johnwparent committed:
review

johnwparent commented:
@scheibelp all requests addressed, please re-review at your convenience.

johnwparent commented:
@spackbot fix style

danlipsa committed:
[@spackbot] updating style on behalf of danlipsa

haampie commented:
@scheibelp can you please do a rebuild everything next time? It breaks other things.
